### PR TITLE
#10056 Fix relative url as basePath

### LINKED ISSUE #2731: [BUG] [typescript-angular] Base Path can't be set to '<empty string>'
#### Bug Report Checklist

- [x] Have you provided a full/minimal spec to reproduce the issue?
- [x] Have you validated the input using an OpenAPI validator ([example](https://apidevtools.org/swagger-parser/online/))?
- [x] What's the version of OpenAPI Generator used?
- [x] Have you search for related issues/PRs?
- [x] What's the actual output vs expected output?
- [x] [Optional] Bounty to sponsor the fix ([example](https://www.bountysource.com/issues/66123212-javascript-client-produces-a-wrong-object-for-a-string-enum-type-that-is-used-with-ref))

##### Description

If the `basePath` in the config is set to `''` (empty string). The configuration is ignored and `'http://localhost'` is used instead. That is unexpected behaviour. 

###### Background
The basePath of the typescript-angular generated Files are currently set to `'http://localhost'` (as a Fallback). When trying to setup the basePath to `''` (empty String) the config is not used as the generated line of code `configuration.basePath || basePath || this.basePath;` is evaluated to false, because empty string in javascript is false. 

```typescript
  protected basePath = 'http://localhost';
  public configuration = new Configuration();

  constructor(
    protected httpClient: HttpClient,
    @Optional() @Inject(BASE_PATH) basePath: string,
    @Optional() configuration: Configuration
  ) {
    if (configuration) {
      this.configuration = configuration;
      this.configuration.basePath = configuration.basePath || basePath || this.basePath;
    } else {
      this.configuration.basePath = basePath || this.basePath;
    }
  }
```

##### openapi-generator version

4.0.0-beta2

##### Command line used for generation

```bash
npm install @openapitools/openapi-generator-cli@0.0.8-4.0.0-beta2
openapi-generator generate -i swagger.json -g typescript-angular -o projects/client-api/src/lib/ --additional-properties="ngVersion=6.1.7" --additional-properties="supportsES6=true"
```

##### Steps to reproduce

1. Generate typescript-angular package from openapi.config
2. Configure the `ApiModule` in app.module.ts
3. Use a service from the generated typescript-angular package. 

##### Related issues/PRs

Nothing found.

##### Suggest a fix
Two ways:
1. Remove the fallback basePath by setting it to `''` (empty string)
2. Fix the evalutation `basePath` property.

### LINKED ISSUE #8576: [BUG][typescript-fetch] Base Path can't be set to relative path
#### Bug Report Checklist

- [ ] Have you provided a full/minimal spec to reproduce the issue?
- [ ] Have you validated the input using an OpenAPI validator ([example](https://apidevtools.org/swagger-parser/online/))?
- [ ] Have you [tested with the latest master](https://github.com/OpenAPITools/openapi-generator/wiki/FAQ#how-to-test-with-the-latest-master-of-openapi-generator) to confirm the issue still exists?
- [ ] Have you searched for related issues/PRs?
- [ ] What's the actual output vs expected output?
- [ ] [Optional] Sponsorship to speed up the bug fix or feature request ([example](https://github.com/OpenAPITools/openapi-generator/issues/6178))

##### Description

When the server url is set to a relative path, in the generated source files it will be prefixed with "https://localhost"

##### openapi-generator version

latest

##### OpenAPI declaration file content or url

```yaml
"servers": [
    {
      "url": "/my-api"
    }
  ]
```

##### Generation Details

openapi-generator-cli generate --skip-validate-spec --input-spec ./my-api.json --generator-name typescript-fetch --output gen/my-api --config my-api-config.json

my-api-config.json
```json
{
  "npmName": "my-api",
  "npmVersion": "1.0.0",
  "snapshot": false,
  "supportsES6": false,
  "modelPropertyNaming": "original"
}
```

##### Steps to reproduce

Run the above command on a yaml file with server url set to a relative path

##### Related issues/PRs

#2731
#2731

### LINKED ISSUE #10056: [BUG][JAVA] Bug the getServerURL method doesnt respect relative urls as described in the openapi definition
#### Bug Report Checklist

- [x] Have you provided a full/minimal spec to reproduce the issue?
- [x] Have you validated the input using an OpenAPI validator ([example](https://apidevtools.org/swagger-parser/online/))?
- [x] Have you [tested with the latest master](https://github.com/OpenAPITools/openapi-generator/wiki/FAQ#how-to-test-with-the-latest-master-of-openapi-generator) to confirm the issue still exists?
- [x] Have you searched for related issues/PRs?
- [x] What's the actual output vs expected output?
- [ ] [Optional] Sponsorship to speed up the bug fix or feature request ([example](https://github.com/OpenAPITools/openapi-generator/issues/6178))

##### Description

The generator prefixes the basePath with http://localhost when servers[0].url == '/api'
resulting in a basePath == 'http://localhost/api'
if the generated client is used in multiple domains the basePath should be set relative to make the generated client more flexible i think?? 

Also the openapi specs explicitly state that the url may be relative which isnt the possible with this generator at the moment
[4.8.5.1 Fixed Fields](https://spec.openapis.org/oas/latest.html#fixed-fields-3)
> REQUIRED. A URL to the target host. This URL supports Server Variables and MAY be relative, to indicate that the host location is relative to the location where the OpenAPI document is being served. Variable substitutions will be made when a variable is named in {brackets}.

##### openapi-generator version

5.2 and master

##### OpenAPI declaration file content or url

This is an example `spec.yaml`

```yaml
openapi: "3.0.3"
info: 
  title: "test api"
  description: "unofficial api test"
  termsOfService: "https://localhost/api/"
  contact: 
    name: "API Support"
    url: "https://localhost/helpdesk/"
    email: "info@localhost"
  license: 
    name: MIT
  version: v1
servers: 
  - url: "/api"
tags: 
  - name: account
paths:
  "/account/v1/json/login"
    post: 
      description: ""
      tags: 
        - account
      operationId: "account.login"
      responses: 
        default: 
          description: ""
          content: 
            "application/json": 
              schema: 
                "$ref": "#/components/schemas/account.login.response.default"
      requestBody: 
        description: ""
        content: 
          "application/json": 
            schema: 
              type: object
              properties: 
                emailAddress: 
                  type: string
                password: 
                  type: string
                  
```

##### Steps to reproduce

1. save the example file as `api.yaml`
2. run `./openapi-generator-cli.jar generate -g typescript-axios -i api.yaml -o client`
3. the generated client library added http://localhost to the relative url

##### Suggest a fix

https://github.com/OpenAPITools/openapi-generator/blob/870ef3a87ac5a9751198ea0f0e36c821563b5e61/modules/openapi-generator/src/main/java/org/openapitools/codegen/utils/URLPathUtils.java#L69

### LINKED ISSUE #10697: [BUG] [typescript-angular] Use a relative basePath
#### Bug Report Checklist

- [x] Have you provided a full/minimal spec to reproduce the issue?
- [x] Have you validated the input using an OpenAPI validator ([example](https://apidevtools.org/swagger-parser/online/))?
- [x] Have you [tested with the latest master](https://github.com/OpenAPITools/openapi-generator/wiki/FAQ#how-to-test-with-the-latest-master-of-openapi-generator) to confirm the issue still exists?
- [x] Have you searched for related issues/PRs?
- [x] What's the actual output vs expected output?
- [ ] [Optional] Sponsorship to speed up the bug fix or feature request ([example](https://github.com/OpenAPITools/openapi-generator/issues/6178))

##### Description

I am using the typescript-angular generator. In the api.service it generates a basePath using http://localhost when you use a relative path.
```
protected basePath = 'http://localhost';
```

##### openapi-generator version

5.3.1

##### OpenAPI declaration file content or url

```
servers:
  - url: '/api'
```

##### Steps to reproduce

Set the url in the openapi specification to a relative path, for example `/api`. When you run the typescript-angular generator now, it will generate the basePath in the api.service as `http://localhost/api` instead of `/api`

##### Related issues/PRs

#2731
#8576

##### Suggest a fix

Remove the prefixed `http://localhost` when a relative path is used.

### PR DESCRIPTION
**Pull request by code1x1**
fixes #2731
fixes #8576
fixes #10697
fixes #10056

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `5.3.x`, `6.0.x`
- [ ] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

### CODE REVIEW

**Comment by code1x1** [comment]
Push

**Comment by agilob** [comment]
Can you add some tests for it?

**Comment by code1x1** [comment]
@agilob sorry for the late response. just added a test as you requested

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/utils/URLPathUtils.java [hunk 1]
```diff
@@ -232,4 +232,12 @@ private static URL getDefaultUrl() {
             return null;
         }
     }
+
+    public static boolean isRelativeUrl(List<Server> servers) {
```
Comment (agilob) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/utils/URLPathUtils.java:236]: replace tabs with spaces

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/utils/URLPathUtils.java [hunk 2]
```diff
@@ -232,4 +232,12 @@ private static URL getDefaultUrl() {
...
+		if (servers.size() > 0) {
```
Comment (agilob) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/utils/URLPathUtils.java:237]: can `servers` be null or empty?

File: modules/openapi-generator/src/test/java/org/openapitools/codegen/DefaultGeneratorTest.java [hunk 3]
```diff
@@ -662,6 +662,30 @@ public void testHandlesTrailingSlashInServers() {
         Assert.assertEquals(servers.get(1).url, "http://trailingshlash.io:80/v1");
         Assert.assertEquals(servers.get(2).url, "http://notrailingslash.io:80/v2");
     }
+    
+    @Test
+    public void testHandlesRelativeUrlsInServers() {
```
Comment (agilob) [inline modules/openapi-generator/src/test/java/org/openapitools/codegen/DefaultGeneratorTest.java:667]: method `configureGeneratorProperties` is package protected, you didn't need to write this whole test with setup of new yaml file

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/utils/URLPathUtils.java (hunk 1 above)
Comment (agilob) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/utils/URLPathUtils.java:236]: why is it `public static`?

**Comment by agilob** [comment]
just a note: You didn't need to create new yaml to test `configureGeneratorProperties`. This method is package-private so can be tested directly.

**Comment by code1x1** [comment]
> just a note: You didn't need to create new yaml to test `configureGeneratorProperties`. This method is package-private so can be tested directly.

ok so how would you write this test? i just adopted the test closest to my issue `testHandlesTrailingSlashInServers` and changed it to fit my needs :sweat_smile:

**Comment by code1x1** [comment]
Are there any open tasks?

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultGenerator.java [hunk 4]
```diff
@@ -268,7 +268,11 @@ void configureGeneratorProperties() {
         URL url = URLPathUtils.getServerURL(openAPI, config.serverVariableOverrides());
         contextPath = removeTrailingSlash(config.escapeText(url.getPath())); // for backward compatibility
         basePathWithoutHost = contextPath;
-        basePath = removeTrailingSlash(config.escapeText(URLPathUtils.getHost(openAPI, config.serverVariableOverrides())));
+        if (URLPathUtils.isRelativeUrl(openAPI.getServers())) {
+        	basePath = removeTrailingSlash(basePathWithoutHost);
```
Comment (agilob) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultGenerator.java:272]: this is still formatted with tabs

**Comment by code1x1** [comment]
can we close the pull request?
my last 2 commits (2e879c4, 05c6852) resolved the tabs issue

**Comment by surban** [comment]
When the server URL consists of a single slash, i.e. `/`, it is erroneously treated as absolute and replaced with `http://localhost`.

**Pull request merged**